Change Log
===========

2.6.12
------
* Parse ``fastq-dump`` version 3.*

2.6.11
-------
* Fix parsing of ``fastq-dump`` version.

2.6.10
-------
* Pin ``numpy`` to < 1.2 for this: https://github.com/numpy/numpy/issues/18355

2.6.9
-----
* Pin ``pystan`` version to < 3.

2.6.8
------
* Fix issue with parsing ``fastq-dump`` version.

2.6.7
-----
* Pin ``matplotlib``, ``pandas``, and ``plotnine`` versions to avoid errors, and stop testing ``barcodeInfoToCodonVariantTable`` as ``dms_variants`` has incompatible versions.

2.6.6
-----
* Ensure that ``dms2_batch_bcsubamp`` propagates ``--bclen`` / ``--bclen2`` values of 0.

2.6.5
-----
* Fix bug in `fastqFromSRA` for newer versions of ``fastq-dump``.

* Specify `pandas` < 1.0

2.6.4
------
* Added `--bcinfo_csv` option to `dms2_bcsubamp`.

2.6.3
------
* Edited `syn_selection` to add pseudocounts for odds ratio calculations in order to avoid null/inf values.

2.6.2
------
* Added 'syn_selection'.

2.6.1
------
* Added `utils.codon_to_nt_counts`.

2.6.0
------
* Support python3.7 and `pystan` >= 2.19 and removed limit that `scipy < 1.3`.

2.5.1
--------
* Remove `barcodes.fracIdentWithinBarcode` as it is **not** correct. There is a new function being added in `alignparse` that calculates the empirical accuracy correctly.

* Remove dependency on `HTSeq`.

* Added `--no_corr` and `--no_avg` option to `dms2_batch_prefs`

2.5.0
------
* Remove `codonvarianttable` module as it is now in `dms_variants <https://github.com/jbloomlab/dms_variants>`_ package

* Remove `neutcurve` module as it is replated by `neutcurve <https://jbloomlab.github.io/neutcurve>`_ package

* Parse - / + signs in PAF alignment scores.

* Works with `pandas` 0.25

2.4.16
------
* Added `diffsel.beta_diversity`

2.4.15
---------
* Relax dependency requirements for `pysam`

2.4.14
-------
* Allow non-integer `--letterheight` for `dms2_logoplot`

2.4.13
------
* Added option for single-color logoplots for `dms2_logoplot`

2.4.12
---------
* Makes sample names a category to preserve order in `plot.plotCodonMutTypes`.

2.4.11
---------
* Adds functionality that can convert the `*bc_info*` files generated by `dms2_bcsubamp` into a `CodonVariantTable`.

2.4.10
-----------
* Added `mle` method to `findSigSel`.

2.4.9
---------
* Allow multiprocessing `fastqFromSRA`

* Updates to genotype-phenotype maps generated from `CodonVariantTable` func scores.

2.4.8
---------
* Allow bottleneck of `None` in `simulateSampleCounts`.

* added `CodonVariantTable.classifyVariants` method

* `dms2_batch_bcsubamp` handles `--bcinfo` argument

2.4.7
--------
* Added `CodonVariantTable.from_simulation` method.

* Handle large facet plots in ``dms2_batch_bcsubamp`` & ``plotSiteDiffSel``

* Correctly handle missing output for individual samples in ``dms2_batch_*`` programs.

2.4.6
----------
* Added function to create `gpmap.GenotypePhenotypeMap` from `CodonVariantTable`

* Don't test `rplot` module, we now recommend `dmslogo <https://jbloomlab.github.io/dmslogo/>`_ instead.

* Work with version >=1.0 of `umi_tools`

2.4.5
----------
* Handle single sample groups in `dms2_batch_diffsel` and `dms2_batch_fracsurvive`

* added `utils.sigFigStr`

* make compatible with latest `natsorted`

* added `diffsel.df_read_filecols`

2.4.4
----------
* Fix to work with pandas 0.24 and newest regex

2.4.3
----------
* Bug fix to plotting mutation frequencies in `CodonVariantTable`

2.4.2
----------
* Enhancements to `codonvarianttable`, including simulating data with simulated latent / global phenotypes, and more plotting.

2.4.1
----------
* Moved `CodonVariantTable` class to its own module, `codonvarianttable`.

2.4.0
-----------
* added `muteffects` data type to ``dms2_logoplot``

* added `highlighted_sites` option to `plot.plotSiteDiffSel`

* added `title` to `siteSubsetGGSeqLogo`

* Improve matching of termini tags when there are mismatches in `dms_tools2.pacbio`

* Some updates to `dms_tools2.minimap2.Mutations` objects.

* Added functionality to `dms_tools2.barcodes` to get identity within barcodes, take consensus of sequences with same barcode, and then use barcoded codon variants to call mutations.

* `pacbio.CCS` handles FASTQ as well as BAM input.

2.3.0
--------
* Require Python 3.6 or greater.

* added `pacbio`, `minimap2`, and `seqnumbering` modules.

* ``import dms_tools2`` now imports all non-optional modules in the package

2.2.9
-------
* Improved appearance of histograms created by `findSigSel`

* Fixed ymin for `dms2_logoplot` with small-valued fracsurvive

2.2.8
--------
* Cleaned up organization of examples and added `Dingens2018` and `Lee2018`.

* Fix setting of `matplotlib` backend, which was causing errors.

* Use loose versions in `sra`.

2.2.7
-------
* Make `diffsel` and `fracsurvive` compatible with `this pandas 0.23 update <https://pandas.pydata.org/pandas-docs/stable/whatsnew.html#assign-accepts-dependent-arguments>`_

2.2.6
-------
* fix bug in `diffsel.tidyToWide` such that sites with matching values (i.e. all 0 values) are not dropped during de-duplication.

* added `utils.codonEvolAccessibility` function


2.2.5
-------
* fix bug in stringency re-scaling by ``dms2_logoplot``

* fix bug in axis scaling in `plot.plotCumulMutCounts`

* added `Haddox2018` and `Doud2018` examples and doc links

2.2.4
----------
* added `contour` option to `plot.plotCorrMatrix`

* added `white_bg` to `plot.plotSiteDiffSel`

2.2.3
------------
* added `plot.findSigSel`

2.2.2
----------
* added `--bclen2` option to ``dms2_bcsubamp``

2.2.1
---------
* added `protstruct` module for operations related to protein structurs

* added `neutcurve` module to fit neutralization curves

* updated required versions of some dependencies

2.2.0
---------
* added `compareprefs` module

* added `rplot` module to enable plotting with `ggseqlogo <https://omarwagih.github.io/ggseqlogo/>`_

* Added `omega` overlay option to ``dms2_logoplot``

* Fix bug with ``dms2_logoplot`` when using wildtype sequence overlays

* Fix bug with ``--fracsurvivemax 0`` to ``dms2_logoplot``

* Scale pseudocounts when using ``dms2_prefs`` with ``--method ratio``, which should give more accurate values when depths differ across samples.

* Some minor bug fixes.

* Fix bug with handling of disulfide-bonded cysteines in ``dssp`` output.

* Added `colors` option to `plot.plotCorrMatrix`

* Added ``--diffprefs`` option to ``dms2_logoplot``

2.1.0
------
* Added programs and docs for `fracsurvive`.

* Added ``--scalebar`` to ``dms2_logoplot``.

* Add `grouplabel` option and preserve group order for faceted plots by batch programs.

* Handle dependencies without `__version__` attribute

2.0.2
------
* Added ``--sitemask`` option to ``dms2_bcsubamp`` / ``dms2_batch_bcsubamp``.

* Standardized color scheme in ``*_cumulmutcounts.pdf`` plot.

* Ensure naturally sorted average prefs from ``dms2_batch_prefs``.

2.0.1
------
* A few packaging changes for PyPI

2.0.0
--------
This version is a complete re-write of `dms_tools <https://github.com/jbloomlab/dms_tools>`_ version 1.2.2.
